Unify invoking JIT operands

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58753 Fix race condition in TP agent
* #58675 Make Future always extract DataPtrs in RPC tests
* #56863 Ensure async_execution works with CUDAFuture
* #57355 Avoid re-doing CUDA stream sync in OwnerRRef
* #58428 Make TP agent use streams from Future when sending response
* #58674 Set and propagate devices in RRef completion future
* #58427 Set streams when invoking UDFs
* #58426 Create CUDA-aware futures in RequestCallback
* #58425 Provide pre-extracted DataPtrs when completing a Future with a Message
* #58424 Allow Future::then to return pre-extracted DataPtrs
* #58423 Always use intrusive_ptr for Message (2 out of 2)
* #58422 Always use intrusive_ptr for Message (1 out of 2)
* #58421 Prevent using anything other than intrusive_ptr for Future
* #58420 Migrate remaining shared_ptr<Future> to intrusive_ptr
* #57861 Make remaining autograd methods return futures
* #57860 Make remaining RRef methods return futures
* #57859 Unify fetching RRefs
* #57858 Deduplicate Python object serialization
* #57857 Simplify process(Script|Python)(Remote)?Call
* #57856 Unify assignment of OwnerRRef result
* #57855 Make processPythonExecution return a future
* #57854 Remove getScriptRemoteCallType
* #57853 Simplify OwnerRRef completion
* #57852 Unify async execution for JIT functions
* #57851 Unify invoking JIT functions
* **#57850 Unify invoking JIT operands**
* #57849 Make some methods of RequestCallback return void instead of bool
* #57848 Centralize setting messageId in RequestCallback
* #57847 Make RequestCallback collect Futures from methods, rather than providing them
* #57846 Add helpers to manipulate futures
* #57638 Reduce overhead when Future invokes callbacks inline
* #57637 Introduce thenAsync method on Future

What I envision is a modular decomposed code, with separate steps which each consume/produce Futures, and which can be chained together to obtain the desired results. One common "starting point" for these chains is the execution of a remote function (Python or JIT or otherwise). I'm thus creating a helper function for one of these, the JIT operators (by deduplicating the places where we used to run them). More will follow.

This deduplication will also help to add CUDA support to JIT RPC, since the execution of the JIT function/operators is where we need to set our custom streams.

Differential Revision: [D28253847](https://our.internmc.facebook.com/intern/diff/D28253847/)

lw committed:
Unify invoking JIT operands

What I envision is a modular decomposed code, with separate steps which each consume/produce Futures, and which can be chained together to obtain the desired results. One common "starting point" for these chains is the execution of a remote function (Python or JIT or otherwise). I'm thus creating a helper function for one of these, the JIT operators (by deduplicating the places where we used to run them). More will follow.

This deduplication will also help to add CUDA support to JIT RPC, since the execution of the JIT function/operators is where we need to set our custom streams.

Differential Revision: [D28253847](https://our.internmc.facebook.com/intern/diff/D28253847/)

[ghstack-poisoned]

lw committed:
Update on "Unify invoking JIT operands"


What I envision is a modular decomposed code, with separate steps which each consume/produce Futures, and which can be chained together to obtain the desired results. One common "starting point" for these chains is the execution of a remote function (Python or JIT or otherwise). I'm thus creating a helper function for one of these, the JIT operators (by deduplicating the places where we used to run them). More will follow.

This deduplication will also help to add CUDA support to JIT RPC, since the execution of the JIT function/operators is where we need to set our custom streams.

Differential Revision: [D28253847](https://our.internmc.facebook.com/intern/diff/D28253847/)

[ghstack-poisoned]

lw committed:
Update on "Unify invoking JIT operands"


What I envision is a modular decomposed code, with separate steps which each consume/produce Futures, and which can be chained together to obtain the desired results. One common "starting point" for these chains is the execution of a remote function (Python or JIT or otherwise). I'm thus creating a helper function for one of these, the JIT operators (by deduplicating the places where we used to run them). More will follow.

This deduplication will also help to add CUDA support to JIT RPC, since the execution of the JIT function/operators is where we need to set our custom streams.

Differential Revision: [D28253847](https://our.internmc.facebook.com/intern/diff/D28253847/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/request_callback_no_python.cpp:
```diff
@@ -661,6 +649,24 @@ bool RequestCallbackNoPython::cudaAvailable() const {
+  }
+  TORCH_INTERNAL_ASSERT(
+      stack.size() == 1,
+      "Return value of a builtin operator or a "
+      "TorchScript function should be a single IValue, got a vector of "
+      "size ",
```
nit: can fit in two lines.

mrshenli reviewed (approved):
LGTM

mrshenli commented on torch/csrc/distributed/rpc/request_callback_no_python.h:
```diff
@@ -124,6 +124,12 @@ class TORCH_API RequestCallbackNoPython : public RequestCallback {
 
+  // Helpers to run user-defined functions, operators and other computations.
+
+  c10::intrusive_ptr<JitFuture> runJitOperator(
```
If this is not used in derived classes, this does not needs to be a member function?

lw replied:
Later in the stack it will be used by `RequestCallbackImpl` (the subclass)

lw committed:
Update on "Unify invoking JIT operands"


What I envision is a modular decomposed code, with separate steps which each consume/produce Futures, and which can be chained together to obtain the desired results. One common "starting point" for these chains is the execution of a remote function (Python or JIT or otherwise). I'm thus creating a helper function for one of these, the JIT operators (by deduplicating the places where we used to run them). More will follow.

This deduplication will also help to add CUDA support to JIT RPC, since the execution of the JIT function/operators is where we need to set our custom streams.

Differential Revision: [D28253847](https://our.internmc.facebook.com/intern/diff/D28253847/)

[ghstack-poisoned]

lw committed:
Update on "Unify invoking JIT operands"


What I envision is a modular decomposed code, with separate steps which each consume/produce Futures, and which can be chained together to obtain the desired results. One common "starting point" for these chains is the execution of a remote function (Python or JIT or otherwise). I'm thus creating a helper function for one of these, the JIT operators (by deduplicating the places where we used to run them). More will follow.

This deduplication will also help to add CUDA support to JIT RPC, since the execution of the JIT function/operators is where we need to set our custom streams.

Differential Revision: [D28253847](https://our.internmc.facebook.com/intern/diff/D28253847/)

[ghstack-poisoned]

lw committed:
Update on "Unify invoking JIT operands"


What I envision is a modular decomposed code, with separate steps which each consume/produce Futures, and which can be chained together to obtain the desired results. One common "starting point" for these chains is the execution of a remote function (Python or JIT or otherwise). I'm thus creating a helper function for one of these, the JIT operators (by deduplicating the places where we used to run them). More will follow.

This deduplication will also help to add CUDA support to JIT RPC, since the execution of the JIT function/operators is where we need to set our custom streams.

Differential Revision: [D28253847](https://our.internmc.facebook.com/intern/diff/D28253847/)

[ghstack-poisoned]

lw committed:
Update on "Unify invoking JIT operands"


What I envision is a modular decomposed code, with separate steps which each consume/produce Futures, and which can be chained together to obtain the desired results. One common "starting point" for these chains is the execution of a remote function (Python or JIT or otherwise). I'm thus creating a helper function for one of these, the JIT operators (by deduplicating the places where we used to run them). More will follow.

This deduplication will also help to add CUDA support to JIT RPC, since the execution of the JIT function/operators is where we need to set our custom streams.

Differential Revision: [D28253847](https://our.internmc.facebook.com/intern/diff/D28253847/)

[ghstack-poisoned]
